feat:  Letter Management feature standardize official correspondence.

aadhilpm opened the pull request:
This PR introduces a Letter Management, allowing users to create consistent, professional, and reusable official letters directly from ERPNext.

Since the company issues many letters regularly, it is important to maintain a proper record. This ensures that all letters released from the company can be tracked, referenced, and retrieved when required.


https://github.com/user-attachments/assets/c2ae5125-8716-4b43-867b-84ff2d02ff06


**Features included**

- Letter Type master for categorizing letter formats
- Letter Template (Jinja) with dynamic subject and content placeholders
- Letter (Submittable) document with: Dynamic recipient (Customer / Supplier / Employee / Shareholder / Contact)
- Template rendering into letter content
- Printing and layout
- Letterhead selection based on Company defaults
- Print language support

**Benefits**
Standardized and consistent letter formats across all departments
Quick creation and printing without manual formatting
Centralized tracking of all letters released by the company

**Usage**

1. Create a Letter Type
2. Create a Jinja Letter Template
3. Create a Letter and select recipient and template
4. Print using the Letter Standard Print Format

<img width="4112" height="3047" alt="letterscreenshot" src="https://github.com/user-attachments/assets/1bdcbec7-48e3-4741-a3b7-ea4a7dd0f1bd" />

Docs: https://docs.frappe.io/erpnext/letter

aadhilpm committed:
feat:  Letter Management feature standardize official correspondence.
- Letter Type master for categorizing letter formats
- Letter Template (Jinja) with subject and content placeholders
- Letter (Submittable) with dynamic recipient (Customer / Supplier / Employee / Shareholder / Contact)
- Auto-fetch recipient name and address
- Company and recipient address display formatting
- Letter template rendering into letter content
- Letterhead selection from company defaults
- Minimal professional print format (letterhead + footer)

This makes creating, formatting and printing official letters fast and consistent, without manual formatting each time.

aadhilpm committed:
fix: linter and prettier formatting issues

aadhilpm committed:
fix: add permission and error handling

aadhilpm committed:
fix: apply ruff formatter

aadhilpm committed:
fix:  permission checks and type hints

aadhilpm committed:
chore: trigger CI

aadhilpm committed:
chore: trigger CI

aadhilpm commented:
> **Actionable comments posted: 0**
> 
> 🧹 Nitpick comments (1)
> > erpnext/crm/doctype/letter/letter.py (1)> `71-80`: **Consider extracting the recipient name field logic to reduce duplication.**
> > > The logic for determining `name_field` and retrieving `recipient_name` (lines 71-80) duplicates the implementation in the `Letter` class methods (`get_recipient_name_field` and `set_recipient_name`, lines 48-54 and 40-46). This creates a maintenance burden—any changes to the name field mapping would need to be updated in both places.
> > > Consider extracting this logic into a module-level helper function that both the class and the whitelisted function can call:
> > > ```diff
> > > +def _get_recipient_name_field(recipient_type: str) -> str:
> > > +	"""Get the appropriate name field for a given recipient type."""
> > > +	if recipient_type == "Shareholder":
> > > +		return "title"
> > > +	elif recipient_type == "Contact":
> > > +		return "full_name"
> > > +	else:
> > > +		return recipient_type.lower() + "_name"
> > > +
> > > +
> > >  class Letter(Document):
> > >  	# ...
> > >  
> > >  	def get_recipient_name_field(self):
> > > -		if self.recipient_type == "Shareholder":
> > > -			return "title"
> > > -		elif self.recipient_type == "Contact":
> > > -			return "full_name"
> > > -		else:
> > > -			return self.recipient_type.lower() + "_name"
> > > +		return _get_recipient_name_field(self.recipient_type)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Then update `get_recipient_details` to use the helper:
> > > ```diff
> > > -	name_field = (
> > > -		"title"
> > > -		if recipient_type == "Shareholder"
> > > -		else ("full_name" if recipient_type == "Contact" else recipient_type.lower() + "_name")
> > > -	)
> > > +	name_field = _get_recipient_name_field(recipient_type)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > This improves maintainability and readability by eliminating the nested ternary expression.
> 
> 📜 Review details

I followed the same approach used in Payment Entry. In Payment Entry, the party_name is fetched using this logic, and I’ve applied a similar concept to fetch recipient_name.

https://github.com/frappe/erpnext/blob/eb38accb336f268676ce41cc4fc6d53a8ba2b183/erpnext/accounts/doctype/payment_entry/payment_entry.py#L2752

aadhilpm committed:
fix: improve UX by removing msgprint from set_query

aadhilpm committed:
fix: trim trailing whitespace

diptanilsaha commented:
@aadhilpm Currently, we are not adding any new features to ERPNext's CRM Module. Check out [Frappe CRM](https://github.com/frappe/crm). 

Also, I recommend opening a GitHub issue for a feature request first and then working on it.

aadhilpm commented:
@diptanilsaha 

Thanks for the update.

Can you please suggest which module this feature would be suitable for? I added it under this module because DocTypes like Contract, Appointment, etc. already belong here. If there is a better place, I will update the code accordingly.

diptanilsaha commented:
@aadhilpm you can add a feature request on Frappe CRM.

aadhilpm committed:
refactor: move Letter Management from CRM to Utilities module

aadhilpm committed:
test: add tests for Letter and Letter Type doctype

aadhilpm commented:
@diptanilsaha  Since Letters are needed across the company, I moved the feature from CRM to Utilities and updated all the code. Let me know if anything else needs to be changed.

aadhilpm commented:
@diptanilsaha Since Letters are needed across the company, I moved the feature from CRM to Utilities and updated all the code. Let me know if anything else needs to be changed.

aadhilpm commented:
@diptanilsaha

aadhilpm commented:
Active

aadhilpm commented:
Active

aadhilpm commented:
@diptanilsaha

aadhilpm commented:
Active
